### ISSUE TITLE
java -jar /home/shlomif/Download/unpack/net/www/validator/build/dist/vnu.jar --format json --Werror --skip-non-html /tmp/zjzj generates invalid JSON

### ISSUE THREAD

**Issue by shlomif** [validator/validator#722]
Hi!

with the contents of http://www.shlomifish.org/Files/files/arcs/shlomif-homepage-for-vnu--v0.tar.xz as well as validator latest trunk (as built on https://www.mageia.org/en/ v7 x86-64) the command

```
shlomif[homepage]:$trunk$ java -jar /home/shlomif/Download/unpack/net/www/validator/build/dist/vnu.jar --format json --Werror --skip-non-html /tmp/zjzj 2>f.json

shlomif[homepage]:$trunk$ gvim f.json 
shlomif[homepage]:$trunk$ json_pp < f.json 
, or } expected while parsing object/hash, at character offset 8571 (before "messages":[]}\n") at /usr/bin/json_pp line 43.
shlomif[homepage]:$trunk$ 

```

please look into fixing it.

**Comment by sideshowbarker** [comment]
I can reproduce this but I have no idea yet what the cause is. I need to make some time to investigate it in detail

**Comment by shlomif** [comment]
On Mon, 22 Oct 2018 17:48:02 -0700
"Michael[tm] Smith" <notifications@github.com> wrote:

> I can reproduce this but I have no idea yet what the cause is. I need to make
> some time to investigate it in detail
> 

Thanks for the update! I guess it can wait for now. Do you know if there are
other easy-to-parse report formats that work better instead with that input? I
want to implement caching by digests/hashes, see
https://github.com/shlomif/shlomi-fish-homepage/blob/caching-vnu-validator/Tests/validate-html-using-vnu.py
.

Regards,

	Shlomi

-- 
-----------------------------------------------------------------
Shlomi Fish       http://www.shlomifish.org/
Buffy Factoids - http://www.shlomifish.org/humour/bits/facts/Buffy/

Q2: Busy people are unproductive. We are very productive and so we’re never
busy.
    — http://www.shlomifish.org/humour/Star-Trek/We-the-Living-Dead/

Please reply to list if it's a mailing list post - http://shlom.in/reply .

**Comment by shlomif** [comment]
@sideshowbarker : thanks! Let me try. I tried with --format xml as well and it too generated malformed output.

**Comment by shlomif** [comment]
@sideshowbarker : with `--format gnu` I am getting an empty input:

```
shlomif[homepage]:$trunk$ java -jar /home/shlomif/Download/unpack/net/www/validator/build/dist/vnu.jar --format gnu --Werror --skip-non-html /tmp/zjzj         
shlomif[homepage]:$trunk$ 
```

note that `shlomif[homepage]:$trunk$ java -jar /home/shlomif/Download/unpack/net/www/validator/build/dist/vnu.jar --format gnu --Werror --skip-non-html /tmp/zjzj/dest/post-incs/t2/humour` yields some errors.

**Comment by shlomif** [comment]
@sideshowbarker : it is the same tree as before. Invoking it on a subtree of the same tree causes it to report some errors while using `--format json` yields a non empty messages array when being called on the whole tree. using `--verbose` seems to indicate that the documents are being checked but no errors are reported.

**Comment by sideshowbarker** [comment]
Thanks for reporting this and sorry for taking so long to get around to investigating it. I’ve merged a fix for it into the sources in change 1ceb8c7.

I still don’t really understand the cause but the symptom seems to be that when the output exceeds 8KB, some of the output right around byte 8192 gets truncated. 

So I guess that’s caused by that code relying a output-stream class that’s using an 8KB buffer, and… something goes wrong at the point where the buffer fills up.

I worked around it by making the code flush the relevant output stream after emitting each JSON object. I guess that probably is not best for performance, but I doubt it actually causes any observable  performance degrade.

Anyway, in the process of investigating this, I also found an outright bug in the existing code that will also cause invalid JSON output for your case. But I’ve fixed that too and will merge the fix shortly.
